pkg/controllers/mcs/service_export_controller.go

```diff
@@ -0,0 +1,487 @@
+	}
+
+	c.InformerManager.Start(cluster.Name, c.StopChan)
+	synced := c.InformerManager.WaitForCacheSync(cluster.Name, c.StopChan)
+	if synced == nil {
+		klog.Errorf("No informerFactory for cluster %s exist.", cluster.Name)
+		return fmt.Errorf("no informerFactory for cluster %s exist", cluster.Name)
+	}
+	for _, gvr := range gvrTargets {
+		if !synced[gvr] {
+			klog.Errorf("Informer for %s hasn't synced.", gvr)
+			return fmt.Errorf("informer for %s hasn't synced", gvr)
+		}
+	}
+
+	return nil
+}
+
+// getEventHandler return callback function that knows how to handle events from the member cluster.
+func (c *ServiceExportController) getEventHandler(clusterName string) cache.ResourceEventHandler {
+	if value, exists := c.eventHandlers.Load(clusterName); exists {
+		return value.(cache.ResourceEventHandler)
+	}
+
+	eventHandler := informermanager.NewHandlerOnEvents(c.genHandlerAddFunc(clusterName), c.genHandlerUpdateFunc(clusterName),
+		c.genHandlerDeleteFunc(clusterName))
+	c.eventHandlers.Store(clusterName, eventHandler)
+	return eventHandler
+}
+
+func (c *ServiceExportController) genHandlerAddFunc(clusterName string) func(obj interface{}) {
+	return func(obj interface{}) {
+		curObj := obj.(runtime.Object)
+		key, err := keys.FederatedKeyFunc(clusterName, curObj)
+		if err != nil {
+			klog.Warningf("Failed to generate key for obj: %s", curObj.GetObjectKind().GroupVersionKind())
+			return
+		}
+		c.worker.AddRateLimited(key)
+	}
+}
+
+func (c *ServiceExportController) genHandlerUpdateFunc(clusterName string) func(oldObj, newObj interface{}) {
+	return func(oldObj, newObj interface{}) {
+		curObj := newObj.(runtime.Object)
+		if !reflect.DeepEqual(oldObj, newObj) {
+			key, err := keys.FederatedKeyFunc(clusterName, curObj)
+			if err != nil {
+				klog.Warningf("Failed to generate key for obj: %s", curObj.GetObjectKind().GroupVersionKind())
+				return
+			}
+			c.worker.AddRateLimited(key)
+		}
+	}
+}
+
+func (c *ServiceExportController) genHandlerDeleteFunc(clusterName string) func(obj interface{}) {
+	return func(obj interface{}) {
+		if deleted, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+			// This object might be stale but ok for our current usage.
+			obj = deleted.Obj
+			if obj == nil {
+				return
+			}
+		}
+		oldObj := obj.(runtime.Object)
+		key, err := keys.FederatedKeyFunc(clusterName, oldObj)
+		if err != nil {
+			klog.Warningf("Failed to generate key for obj: %s", oldObj.GetObjectKind().GroupVersionKind())
+			return
+		}
+		c.worker.AddRateLimited(key)
+	}
+}
+
+func (c *ServiceExportController) handleServiceExportEvent(fedKey keys.FederatedKey) error {
+	_, err := helper.GetObjectFromCache(c.RESTMapper, c.InformerManager, fedKey)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return deleteWorkWithServiceExportDeleteEvent(c.Client, fedKey)
+		}
+		return err
+	}
+
+	// the calculate may be redundant when serviceexport fist created, but it's necessary when the serviceexport recreate
+	// after deleting.
```

Comment:
```suggestion
	// Even though the EndpointSlice will be synced when dealing with EndpointSlice events, thus the 'report' here may
	// be redundant, but it helps to avoid a corner case:
	// If skip report here, after ServiceExport deletion and re-creation, if no EndpointSlice changes, we didn't get a
	// change to sync.
```